src/vs/base/common/map.ts

```diff
@@ -0,0 +1,147 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+'use strict';
+
+interface Entry<T> {
+	next?: Entry<T>;
+	prev?: Entry<T>;
+	key: string;
+	value: T;
+}
+
+/**
+ * A simple Map<T> that optionally allows to set a limit of entries to store. Once the limit is hit,
+ * the cache will remove the entry that was last recently added.
+ */
+export class LinkedMap<T> {
+	protected map: { [key: string]: Entry<T> };
+	private head: Entry<T>;
+	private tail: Entry<T>;
+	private _size: number;
+
+	constructor(private limit = Number.MAX_VALUE) {
+		this.map = Object.create(null);
+		this._size = 0;
+	}
+
+	public get size(): number {
+		return this._size;
+	}
+
+	public set(key: string, value: T): boolean {
+		if (this.map[key]) {
+			return false; // already present!
+		}
+
+		const entry: Entry<T> = { key, value };
+		this.push(entry);
+
+		if (this._size > this.limit) {
```

Comment:
Soon you will end up removing a single item for each new item which is extra cost. Size the limit isn't so hard I would use a load factor, like if 3/4 of limit are reached remove so many items that you at 2/3 usage.